broadinstitute/single_cell_portal_core
# Adding limited CI job to run against staging Terra orchestration API (SCP-5196)

#### BACKGROUND
Our current integration with Terra is such that SCP only ever points at their production instance.  This was done for a variety of reasons, mostly for the uptime commitments and QA/QC.  However, a few times this has led to a scenario where a breaking API change is released to production, which causes all SCP instances to encounter some level of degraded service, depending on the change/regression.

#### CHANGES
This update adds a "smoke test" that will run a limited CI job (only `fire_cloud_client_test.rb`) against the [staging](https://firecloud-orchestration.dsde-staging.broadinstitute.org) instance of orchestration.  By doing so, we aim to catch regressions _before_ they are released.  Staging is the appropriate instance as it is up consistently and receives updates weekly that are intended to be released to production.  This job will only run on commits to `development` or `master`, but also could be configured to run on a regular cadence (nightly, weekly, etc).  

In addition, this update contains other changes to support this work.  Most notably, `FireCloudClient` now uses keyword parameters instead of positional ones to make instantiating clients simpler.  Additionally, `FireCloudProfile` now can be used to make updates to your Terra profile from the console to aid in testing this work.  The necessary projects/permissions were created outside of this PR directly in the staging Terra instance.

#### MANUAL TESTING
In addition to the manual test instructions below, you can see in the "Checks" portion of this PR that the [CI job](https://github.com/broadinstitute/single_cell_portal_core/actions/runs/5591300236) has run successfully.

1. Pull this branch and enter a Rails console session
2. Register your service account in the staging SAM instance, adding in your email address for `contactEmail`:
```
client = FireCloudClient.new(
  api_root: 'https://firecloud-orchestration.dsde-staging.broadinstitute.org'
)
profile = FireCloudProfile.new(
  contactEmail: "< your email >",                                                                             
  email: client.issuer,
  firstName: "SCP Dev",
  lastName: "Service Account",
  institute: "Broad Institute",
  institutionalProgram: "DSP",
  nonProfitStatus: "true"",
  pi: "N/A",
  programLocationCity: "Cambridge",
  programLocationState: "MA",
  programLocationCountry: "USA",
  title: "N/A"
)
client.set_profile(profile.attributes)
```
3. In a separate terminal, run the smoke test shell script from the project root: 
```
# -d flag will run test outside of Docker
./rails_local_setup.rb && source config/secrets/.source_env.bash
bin/run_orch_smoke_test.sh -d
```

You should see similar output to this (note the logging statement that points to staging orchestration):
```
Run options: --seed 17362

# Running:

SUITE FireCloudClientTest starting
Running smoke test against Terra orchestration API at https://firecloud-orchestration.dsde-staging.broadinstitute.org
fire_cloud_client_test.rb:76 test_firecloud_api_available starting
...
```
